### PR TITLE
BUILD: clean out py2 stuff from npy_3kcompat.h

### PR DESCRIPTION
**Pull request by mattip**
- Remove extraneous documentation page that was not linked in
- Remove all the py2 compat defines and unused functions
- Remove now unneeded `#include "numpy/ndarrayobject.h" but add it where it was missing in `numpy/_core/src/multiarray/array_method.c`
- Fix up some uses of py2-isms

I will add a release note once I get a PR number

### CODE REVIEW

File: numpy/_core/include/numpy/npy_3kcompat.h [hunk 1]
```diff
@@ -15,158 +14,12 @@
...
 
-/*
- * PyInt -> PyLong
- */
-
-
-/*
- * This is a renamed copy of the Python non-limited API function _PyLong_AsInt. It is
- * included here because it is missing from the PyPy API. It completes the PyLong_As*
- * group of functions and can be useful in replacing PyInt_Check.
```
Comment (mattip) [inline numpy/_core/include/numpy/npy_3kcompat.h:37]: PyPy has this function from 7.3.13 released about a year ago

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -4702,7 +4702,7 @@ setup_scalartypes(PyObject *NPY_UNUSED(dict))
     DUAL_INHERIT(CDouble, Complex, ComplexFloating);
     SINGLE_INHERIT(CLongDouble, ComplexFloating);
 
-    DUAL_INHERIT2(String, String, Character);
+    DUAL_INHERIT2(String, Bytes, Character);
```
Comment (mattip) [inline numpy/_core/src/multiarray/multiarraymodule.c:4705]: Subtle: this creates a base type inheritance on `PyBytesType`. The replacement was done previously via `#define PyString_Type PyBytes_Type`

**Comment by mattip** [comment]
I started looking at #12324, but got distracted by all the cruft in this file.

File: numpy/_core/include/numpy/npy_3kcompat.h (hunk 1 above)
Comment (mattip) [reply numpy/_core/include/numpy/npy_3kcompat.h:37]: hehe, and now CPython removed it

File: numpy/_core/include/numpy/npy_3kcompat.h [hunk 3]
```diff
@@ -437,8 +262,8 @@ npy_PyFile_CloseFile(PyObject *file)
     return 0;
 }
 
-
-/* This is a copy of _PyErr_ChainExceptions
+/* This is a copy of _PyErr_ChainExceptions, which
+ *  is no longer exported from Python3.12
```
Comment (seberg) [inline numpy/_core/include/numpy/npy_3kcompat.h:266]: Seems like this is now `_PyErr_ChainExceptions1`, but a different thing.

**Review by seberg — APPROVED** [review]
LGTM, since we don't need it.  Some downstream projects will probably notice, but it should be easy enough to adapt or vendor the functions.

So giving others a chance to have a look maybe, but I think this can go in.

**Review by ngoldbaum — APPROVED** [review]
Nice, I also noticed all this cruft when I vendored the pythoncapi-compat header but didn't have the fortitude to clean it up. Appreciate you taking the time!

**Comment by ngoldbaum** [comment]
If you're reading this because it turns out merging this PR breaks your build for some reason and you don't immediately see how to deal with it (e.g. vendor an old copy of the header), please do let us know in an issue.

**Pull request merged**